### DISCUSSION TITLE
CLOG 1.7 - Dev Journal

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#210] [General]
Focus will be on improved usability as an IDE and work on properties

**Comment by rabbibotton** [comment]
Added clog-plotly plugin
https://www.reddit.com/r/Common_Lisp/comments/wso3oj/clogplotly_d3_based_plotly_charts_and_plotting/

**Reply by aykaramba** [reply]
Whoa.  That is awesome.

**Comment by rabbibotton** [comment]
2D Canvas has been enhanced, still a few areas need some small work. The API changed (1st time in CLOG's history sorry) slightly for properties to use setf to set properties, should have been the case from the start.

**Reply by rabbibotton** [reply]
finished this up, only matrix transforms not implemented partially but that can wait till implemented for WebGL as well

**Comment by rabbibotton** [comment]
Added console application template

**Comment by rabbibotton** [comment]
default directories are the current project
project window only window opened at startup now
save button, now is a save button - shift click to save as...
installing a project that depends on a plugin loads it and its control in to the control palette
if ~/common-lisp  directory does not exist uses the quicklisp project directory (technically uses first registered directory) for new projects. I may change to offer a list instead soon.
no longer tries to access ~/common-lisp for project search if the directory does not exist

**Comment by rabbibotton** [comment]
The script ./run-builder now starts builder with random port and shutdown when the first connection to it (usually automatic) closes.
You can also run the script ./make-exe to create an executable for your platform.

**Comment by rabbibotton** [comment]
Save button saves (no dialog) if name known shift-click to do save as...
Render button renders (no dialog) if name known shift-click to do render as...
Animated button when saving and rendering

**Reply by rabbibotton** [reply]
also added ctl-s/command-s to save source files and events.

**Comment by rabbibotton** [comment]
There is now an auto render checkbox on the clog panels, when set (default) when saving the clog file it also renders to lisp

**Comment by rabbibotton** [comment]
There is a new button in the project window called - Rerender that will rerender all the clog files to lisp in a project.

**Comment by rabbibotton** [comment]
place-text-inside-bottom-of and place-text-inside-top-of for adding text before and after in divs, text areas, etc.

**Comment by rabbibotton** [comment]
Side panels disappear when not on a page,panel or event window.

**Comment by rabbibotton** [comment]
Various fixes for Firefox and Safari. All should work equally well on Chrome, Firefox and Safari now.

**Comment by rabbibotton** [comment]
You can now batch render from the OS command line. It will try and open a browser if can not you still need to connect, but opens builder batch rerenders your project.

The command line script run-render takes one parameter the asdf system name:
./run-render :my-sys

This will allow you to have projects in git with out need to store generated .lisp files from .clog files.

This is the content of run-render
sbcl --eval "(ql:quickload :clog/tools)" --eval "(clog-tools:clog-builder :port 0 :app :batch :project $1)"

**Reply by rabbibotton** [reply]
If I was doing a large project I would use something like this to generate lisp from .clog files with a chrome headless.

I would spawn one process with:
sbcl --eval "(ql:quickload :clog/tools)" --eval "(clog-tools:clog-builder :port 58080 :app :batch :project $1 :start-browser nil)"

and then another with:
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --headless --remote-debugging-port=9222 http://127.0.0.1:58080/builder

when the first process returned kill the second.

**Reply by rabbibotton** [reply]
This works for firefox:

/Applications/Firefox.app/Contents/MacOS/firefox --headless http://127.0.0.1:58080/builder

**Reply by rabbibotton** [reply]
btw that process of using --headless will run on servers with no gpu's etc

**Comment by rabbibotton** [comment]
Directory window added to tool menu and off the project window.

**Reply by rabbibotton** [reply]
asdf systems also has a view dir button.

**Comment by rabbibotton** [comment]
```
(defgeneric request-animation-frame (clog-window)
  (:documentation "Requests the browser to send an on-animation-frame
on the next screen redraw. This event only fires one time per request.
The data parementer of the event function contains the time stamp
to the millisecond."))
```

**Comment by rabbibotton** [comment]
clogframe - first step in native clog apps using webview. Works great on OS X. Will test my Linux instructions later. See the discussion of someone can help with Windows.

This stage is simple - your clog app starts and uses uiop:run-program to launch clogframe instead of a browser. See 01-demo.lisp

**Reply by rabbibotton** [reply]
./make-mac
./make-builder
./builder
Love :)

**Reply by rabbibotton** [reply]
./make-linux  should also work, I just don't have till later tomorrow or Sunday access to a Linux machine to test.

**Reply by rabbibotton** [reply]
Had a chance and make-linux works perfectly :)

**Comment by rabbibotton** [comment]
MDI Panel project added

**Comment by rabbibotton** [comment]
confirm on close to save file for source code editor

**Reply by rabbibotton** [reply]
confirm now also on panels

**Comment by rabbibotton** [comment]
New templates choose from any registered local project directory

**Comment by rabbibotton** [comment]
Thanks to @rse3vau we now have : 

ctl-=  Expand region
opt/alt-m Macroexpand

**Comment by rabbibotton** [comment]
Verification control name not blank and if starts with a ( name is quoted with | so still valid symbol
Parent is now a drop down list

**Reply by rabbibotton** [reply]
Color pickers now work properly

**Comment by rabbibotton** [comment]
Directory View now has basic directory and file functions.

**Comment by rabbibotton** [comment]
when asking to open a file if already open switched to that window instead

**Comment by rabbibotton** [comment]
1.7 Released
